src/rust/engine/src/context.rs

```diff
@@ -244,6 +244,18 @@ impl Core {
         Box::new(local_command_runner)
       };
 
+      // Note that the Docker command runner is only used if the Process sets docker_image. So,
+      // it's safe to always create this command runner.
+      let runner = Box::new(docker::CommandRunner::new(
+        runner,
+        local_runner_store.clone(),
+        executor.clone(),
+        local_execution_root_dir.to_path_buf(),
+        named_caches.clone(),
+        immutable_inputs.clone(),
+        exec_strategy_opts.local_keep_sandboxes,
+      )?);
```

Comment:
@stuhood is backtracking already handled correctly here? I'm not sure if the new ordering here impacts things.

https://github.com/pantsbuild/pants/blob/d2b22d1594803c6a13dfcf6915a02e3b770f91dc/src/rust/engine/src/nodes.rs#L436-L447

Reply:
Yea, this looks right.

The "leaf" `CommandRunner` ends up shared between the cached and uncached stack.